app/forms/DateForm.scala

```diff
@@ -62,6 +62,22 @@ object DateForm extends InputFilters {
     }
   }
 
+  def validateEndDate(formData: DateFormData,
+                      taxYear: Int,
+                      benefitType: BenefitType,
+                      isAgent: Boolean,
+                      startDate: LocalDate)
+                     (implicit messages: Messages): Seq[FormError] = {
+    lazy val emptyDateFieldsValidationValue = emptyDateFieldsValidation(formData, benefitType, datePageName = "endDatePage", isAgent)
+    lazy val invalidDateFormatValidationValue = invalidDateFormatValidation(formData, benefitType, datePageName = "endDatePage", isAgent)
+    lazy val endDateSpecificValidationValue = endDateSpecificValidation(formData.toLocalDate.get, taxYear, benefitType, isAgent, startDate)
+
+    emptyDateFieldsValidationValue match {
+      case _ :: _ => emptyDateFieldsValidationValue
+      case _ => if (invalidDateFormatValidationValue.nonEmpty) invalidDateFormatValidationValue else endDateSpecificValidationValue
+    }
```

Comment:
Why do we use `.nonEmpty` inside the 2nd case, but don't use it for the match that tests the same thing? Isn't `.nonEmpty` clearer than the `_ :: _` case (where one has to remember the 2nd `_` can match `Nil`)? If `_ :: _` is clearer why don't we use it in the 2nd case too?

Here are 2 alternatives which use `.nonEmpty` for both tests:

```scala
if (emptyDateFieldsValidationValue.nonEmpty) {
  emptyDateFieldsValidationValue
} else {
  if (invalidDateFormatValidationValue.nonEmpty) invalidDateFormatValidationValue else startDateSpecificValidationValue
}

(emptyDateFieldsValidationValue.nonEmpty, invalidDateFormatValidationValue.nonEmpty) match {
  case (true, _) => emptyDateFieldsValidationValue
  case (false, true) => invalidDateFormatValidationValue
  case (false, false) => startDateSpecificValidationValue
}
```

Alternatively if we want to keep a match for the first test, maybe it's clearer to check for the empty list first:

```scala
emptyDateFieldsValidationValue match {
  case Nil => if (invalidDateFormatValidationValue.nonEmpty) invalidDateFormatValidationValue else startDateSpecificValidationValue
  case _ => emptyDateFieldsValidationValue
}
```

What do you reckon?

Reply:
Hmmm I think I like the 2 option, with the `match case` for the 3 cases? @georgievh what do you think?

Reply:
The first version will work, but as we have discussed in the past we don't like overusing if else clouses.
Second version needs to run two validations always when if the first one fails we don't need to calculate the second.
The third version is wrong. Does not achieve the behavior.

I am happy the way it is now